File: app/src/main/java/io/github/lee0701/heonot/KOKR/InputMethod.java
package io.github.lee0701.heonot.KOKR;

import android.content.Context;
import android.content.res.Resources;
import android.view.View;
import android.widget.LinearLayout;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import io.github.lee0701.heonot.KOKR.modules.InputMethodModule;
import io.github.lee0701.heonot.KOKR.modules.softkeyboard.SoftKeyboard;
import io.github.lee0701.heonot.HeonOt;

public class InputMethod implements Cloneable {

	private String name = "";

	private List<InputMethodModule> modules;

	public InputMethod(InputMethodModule... modules) {
		this.modules = new ArrayList<>();
		this.modules.addAll(Arrays.asList(modules));
	}

	public void registerListeners(HeonOt parent) {
		for(InputMethodModule module : modules) {
			parent.addListener(module);
			module.addListener(parent);
			for(InputMethodModule listener : modules) {
				if(module != listener) module.addListener(listener);
			}
		}
	}

	public void clearListeners() {
		for(InputMethodModule module : modules) {
			module.clearListeners();
		}
	}

	public void init() {
		for(InputMethodModule module : modules) {
			module.init();
		}
	}

	public View createView(Context context) {
		LinearLayout view = new LinearLayout(context);
		for(InputMethodModule module : modules) {
			if(module instanceof SoftKeyboard) {
				view.addView(((SoftKeyboard) module).createView(context));
			}
		}
		return view;
	}

	public String toJSON(int indentSpaces) throws JSONException {
		JSONObject methodObject = new JSONObject();

		methodObject.put("name", getName());

		JSONArray modules = new JSONArray();

		for(InputMethodModule module : this.modules) {
			modules.put(module.toJSONObject());
		}

		methodObject.put("modules", modules);
		if(indentSpaces > 0) return methodObject.toString(indentSpaces);
		else return methodObject.toString();
	}

	public static InputMethod loadJSON(String methodJson) throws JSONException {
		JSONObject methodObject = new JSONObject(methodJson);

		JSONArray modulesArray = methodObject.optJSONArray("modules");

		List<InputMethodModule> modules = new ArrayList<>();
		if(modulesArray != null) {
			for(int i = 0 ; i < modulesArray.length() ; i++) {
				JSONObject module = modulesArray.getJSONObject(i);
				String name = module.optString("name");
				String className = module.getString("class");
				try {
					Class<?> moduleClass = Class.forName(className);
					InputMethodModule m = (InputMethodModule) moduleClass.getDeclaredConstructor().newInstance();
					m.setName(name);
					JSONArray props = module.optJSONArray("properties");
					if(props != null) {
						for(int j = 0 ; j < props.length() ; j++) {
							JSONObject prop = props.getJSONObject(j);
							String key = prop.optString("key", null);
							Object value = prop.opt("value");
							if(key != null) m.setProperty(key, value);
						}
					}
					modules.add(m);
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
		}

		InputMethodModule[] arr = new InputMethodModule[modules.size()];
		InputMethod method = new InputMethod(modules.toArray(arr));
		method.setName(methodObject.optString("name"));

		return method;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public List<InputMethodModule> getModules() {
		return modules;
	}

	public void setModules(List<InputMethodModule> modules) {
		this.modules = modules;
	}

	@Override
	public Object clone() {
		InputMethodModule[] modules = new InputMethodModule[this.modules.size()];
		for(int i = 0 ; i < modules.length ; i++) {
			modules[i] = (InputMethodModule) this.modules.get(i).clone();
		}
		InputMethod cloned = new InputMethod(modules);
		cloned.setName(getName());
		return cloned;
	}
}

File: heonot-core/src/main/java/io/github/lee0701/heonot/inputmethod/InputMethod.kt
package io.github.lee0701.heonot.inputmethod

import io.github.lee0701.heonot.HeonOt
import io.github.lee0701.heonot.inputmethod.modules.InputMethodModule
import org.greenrobot.eventbus.EventBus
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

open class InputMethod {
    var name: String = ""
    var modules: MutableList<InputMethodModule>

    constructor(modules: MutableList<InputMethodModule>) {
        this.modules = modules
    }

    constructor(original: InputMethod) {
        this.name = original.name
        this.modules = original.modules.map { it.clone() }.toMutableList()
    }

    fun registerListeners() {
        modules.forEach {
            EventBus.getDefault().register(it)
        }
    }

    fun clearListeners() {
        modules.forEach {
            EventBus.getDefault().unregister(it)
        }
    }

    fun init() {
        modules.forEach {
            it.init()
        }
    }

    fun pause() {
        modules.forEach {
            it.pause()
            EventBus.getDefault().unregister(it)
        }
    }

    @Throws(JSONException::class)
    fun toJSON(indentSpaces: Int): String {
        val methodObject = JSONObject()

        methodObject.put("name", name)

        val modules = JSONArray()

        this.modules.forEach {
            modules.put(it.toJSONObject())
        }

        methodObject.put("modules", modules)
        return if (indentSpaces > 0) {
            methodObject.toString(indentSpaces)
        } else {
            methodObject.toString()
        }
    }

    @Deprecated("Use copy constructor instead", replaceWith = ReplaceWith("InputMethod"))
            /**
             * It is deprecated. Use copy constructor instead.
             */
    fun clone(): Any {
        return InputMethod(this)
    }

    companion object {
		@JvmStatic
        @Throws(JSONException::class)
        fun loadJSON(methodJson: String): InputMethod {
            val methodObject = JSONObject(methodJson)

            val modulesArray = methodObject.optJSONArray("modules")

            val modules = arrayListOf<InputMethodModule>()
            if (modulesArray != null) {
                for (i in 0 until modulesArray.length()) {
                    val module = modulesArray.getJSONObject(i)
                    val name = module.optString("name")
                    val className = module.getString("class")
                    try {
                        val moduleClass = run {
							try {
								Class.forName(HeonOt.INSTANCE?.modulePackageName + "." + className)
							} catch(ex: ClassNotFoundException) {
								Class.forName("io.github.lee0701.heonot.inputmethod.modules." + className)
							}
						}
                        val m = moduleClass.getDeclaredConstructor().newInstance() as InputMethodModule
                        m.name = name
                        val props = module.optJSONObject("properties")
                        if (props?.names() != null) {
                            val names = props.names()
                            for (j in 0 until names.length()) {
                                val key = names.getString(j)
                                m.setProperty(key, props.opt(key))
                            }
                        }
                        modules.add(m)
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }

                }
            }

            val method = InputMethod(modules)
            method.name = methodObject.optString("name")

            return method
        }
    }
}
